pep-0567.rst

```diff
@@ -419,35 +424,56 @@ points to a ``_ContextData`` object::
             raise LookupError
 
         def set(self, value):
-            ts : PyThreadState = PyThreadState_Get()
-            data : _ContextData = ts.context_data
+            ts: PyThreadState = PyThreadState_Get()
+            if ts.context is None:
+                ts.context = Context()
 
+            data: _ContextData = ts.context._data
             try:
                 old_value = data.get(self)
             except KeyError:
                 old_value = Token.MISSING
 
-            ts.context_data = data.set(self, value)
-            return Token(self, old_value)
+            updated_data = data.set(self, value)
+            ts.context._data = updated_data
+            return Token(ts.context, self, old_value)
 
         def reset(self, token):
+            if token._var is not self:
+                raise ValueError(
+                    "Token was created by a different ContextVar")
+
+            ts: PyThreadState = PyThreadState_Get()
+            if token._ctx is not ts.context:
+                raise ValueError(
+                    "Token was created in a different Context")
+
             if token._used:
                 return
 
             if token._old_value is Token.MISSING:
-                ts.context_data = data.delete(token._var)
+                ts.context._data = data.delete(token._var)
             else:
-                ts.context_data = data.set(token._var,
-                                           token._old_value)
+                ts.context._data = data.set(token._var,
+                                            token._old_value)
 
             token._used = True
 
+Note that the in the reference implementation, ``ContextVar.get()``
+has an internal cache for the most recent value, which allows to
+bypass a hash lookup.  This is similar to the optimization the
+``decimal`` module implements to retrieve its context from
+``PyThreadState_GetDict()``.  See :pep:`550` which explains the
+implementation of the cache in great detail.
+
+The ``Token`` class is implemented as follows::
 
     class Token:
 
         MISSING = object()
 
-        def __init__(self, var, old_value):
+        def __init__(self, ctx, var, old_value):
```

Comment:
Maybe spell out `context`? (But not `variable`.) My brain keeps stumbling over `ctx` when I speak the code in my head. :-)

Reply:
Changed it to 'context'.